### DISCUSSION TITLE
variables wrapped in before block not accessible in beforeAfter after the first call?

### DISCUSSION THREAD

**Discussion by aarona** [cypress-io/cypress#24727] [Questions and Help]
I don't know if this is a bug or I'm just using Cypress incorrectly so before I assume there's something wrong with Cypress, I'm asking this here in case there's a "Cypress" way of accomplishing what I'm trying to do. I posted this question on SO so If you want to answer it there, you can do that here: https://stackoverflow.com/questions/74483194

Why don't I have access to `@testVariable` after the first `beforeEach`? I should be able to have access to it for each test in this context.

```
describe('Testing before and beforeEach issue', () => {
  context('testing before and beforeEach', () => {
    before(() => {
      cy.wrap("TESTING 123").as("testVariable")
    })

    beforeEach(() => {
      // This fails after the first beforeEach call.
      cy.get("@testVariable").then((testVariable) => {
        cy.log(`THIS IS A TEST: ${testVariable}`)
      })
    })

    // This test is reached and passes
    it('Testing the variable one time', () => {
      cy.log("THIS IS LOG 1")
    })

    // This test should also pass but is not reached before
    // an error occurs in beforeEach
    it('Testing the variable a second time', () => {
      cy.log("THIS IS LOG 2")
    })
  })
})
```
Here is the test output:

```
  Testing before and beforeEach issue
    testing before and beforeEach
      ✓ Testing the variable one time
      1) "before each" hook for "Testing the variable a second time"


  1 passing (684ms)
  1 failing

  1) Testing before and beforeEach issue
       testing before and beforeEach
         "before each" hook for "Testing the variable a second time":
     CypressError: `cy.get()` could not find a registered alias for: `@testVariable`.
You have not aliased anything yet.

Because this error occurred during a `before each` hook we are skipping the remaining tests in the current suite: `testing before and beforeEach`
```

**Comment by portyu9** [comment]
Hi @aarona 

This is expected Cypress behavior: **aliases are reset before every test**.

Your alias is created only once:

```js
before(() => {
  cy.wrap('TESTING 123').as('testVariable')
})
```

It therefore exists for the first test, gets cleared during test cleanup, and is not recreated for the second test.

Create it in `beforeEach()` instead:

```js
beforeEach(() => {
  cy.wrap('TESTING 123').as('testVariable')
})

beforeEach(() => {
  cy.get('@testVariable').then((testVariable) => {
    // use testVariable
  })
})
```

Or combine the setup into one `beforeEach()`.

For a simple immutable string like this you could also use a normal closure variable and avoid an alias altogether, but if you want to use Cypress aliases, `beforeEach()` is the correct lifecycle hook.

Please mark this answer as accepted if it helped, thank you.
